stdlib/Logging/src/ConsoleLogger.jl

```diff
@@ -0,0 +1,112 @@
+
+# Formatting of values in key value pairs
+showvalue(io, msg) = show(io, "text/plain", msg)
+function showvalue(io, e::Tuple{Exception,Any})
+    ex,bt = e
+    showerror(io, ex, bt; backtrace = bt!=nothing)
+end
+showvalue(io, ex::Exception) = showvalue(io, (ex,catch_backtrace()))
+
+function handle_message(logger::ConsoleLogger, level, message, _module, group, id,
+                        filepath, line; maxlog=nothing, kwargs...)
+    if maxlog != nothing && maxlog isa Integer
+        remaining = get!(logger.message_limits, id, maxlog)
+        logger.message_limits[id] = remaining - 1
+        remaining > 0 || return
+    end
+    levelstr = string(level)
+    level != Warn || (levelstr = "Warning")
+    color = level < Info  ? Base.debug_color() :
+            level < Warn  ? Base.info_color() :
+            level < Error ? Base.warn_color() :
+                            Base.error_color()
+    buf = IOBuffer()
+    iob = IOContext(buf, logger.stream)
+    if logger.show_limited
+        iob = IOContext(iob, :limit=>true)
+    end
+    msglines = split(chomp(string(message)), '\n')
+    prefixwithlevel = level >= logger.min_level_prefix
+    dsize = displaysize(logger.stream)
+    width = dsize[2]
+    locationstr = string(levelstr, " @ ", _module, " ", basename(filepath), ":", line)
+    singlelinewidth = 2 + length(msglines[1]) +
+                      (prefixwithlevel ? length(levelstr) + 2 : 0) +
+                      length(locationstr)
+    if length(msglines) + length(kwargs) == 1 && singlelinewidth <= width
+        print_with_color(color, iob, "[ ", bold=true)
+        if prefixwithlevel
+            print_with_color(color, iob, levelstr, ": ", bold=true)
+        end
+        print(iob, msglines[1])
+        print(iob, " "^(width - singlelinewidth))
+    else
+        print_with_color(color, iob, "┌ ", bold=true)
+        if prefixwithlevel
+            print_with_color(color, iob, levelstr, ": ", bold=true)
+        end
+        println(iob, msglines[1])
+        for i in 2:length(msglines)
+            print_with_color(color, iob, "│ ", bold=true)
+            println(iob, msglines[i])
+        end
+        valbuf = IOBuffer()
+        valio = IOContext(IOContext(valbuf, iob),
+                          :displaysize=>(1 + dsize[1]÷(length(kwargs)+1),dsize[2]-5))
+        for (key,val) in pairs(kwargs)
+            print_with_color(color, iob, "│ ", bold=true)
+            print(iob, "  ", key, " =")
+            showvalue(valio, val)
+            vallines = split(String(take!(valbuf)), '\n')
+            if length(vallines) == 1
+                println(iob, " ", vallines[1])
+            else
+                println(iob)
+                for line in vallines
+                    print_with_color(color, iob, "│    ", bold=true)
+                    println(iob, line)
+                end
+            end
+        end
+        print_with_color(color, iob, "└ ", bold=true)
```

Comment:
For consistency with single line mode, `locationstr` could be printed at the last message/kwarg line if it fits.

Reply:
That would be nice. It does complicate the implementation significantly.

Reply:
I think I'll punt on this for now.  Getting right justification correct in general seems to involve removing ANSI escape sequences from the formatted values to determine the text width of the line prefix. It really seems somewhat messy.